### ISSUE TITLE
`scatter` operand reshape to 1-D emits TILE layout - degenerate height 1→32 padding causes 32× DRAM blow-up (OOMs GNN training), even though the next op converts it to row-major

### ISSUE THREAD

**Issue by St1viG** [tenstorrent/tt-mlir#8887]
### Describe the bug

Follow-up to #8714 / #8718. With that fix, stock `GATv2Conv` scatter now **compiles** and runs forward/inference natively on TT - thank you. But full-graph **training** on PubMed now hits a DRAM OOM that traces to a **lowering inefficiency**, not GAT math.

When a scatter source operand of shape `[E, F]` is flattened to 1-D `[E*F]` before the scatter, the reshape produces the 1-D tensor in **TILE** layout. A 1-D tensor is mapped to `[1, E*F]` and tiled into 32×32 tiles, so the **degenerate height 1 pads up to a full tile of 32 rows → a 32× memory blow-up**. The very next op converts that buffer to **row-major** (which is what `ttnn.scatter` actually consumes), so the tile-padded buffer is **pure transient waste** - but its allocation alone overflows DRAM.

For GATv2 on full PubMed the wasted transient is **887 MB** for a logical tensor of only ~28 MB. The op is correct; it just allocates 32× more than needed.

**IR from the compiled training graph** (message-aggregation scatter operand prep):

```mlir
// %120 : per-edge message [E_sl, heads*out]
%121 = "ttnn.reshape"(%120) <{shape = [6935360 : i32]}>
     : (tensor<108365x64xf32, #ttnn_layout46>)
    -> tensor<6935360xf32, #ttnn_layout57>          // 1-D, TILE layout  ← the blow-up
%128 = "ttnn.to_layout"(%121) <{layout = #ttnn.layout<row_major>}>
     : (tensor<6935360xf32, #ttnn_layout57>)
    -> tensor<6935360xf32, #ttnn_layout60>          // converted to ROW-MAJOR (~28 MB)
%129 = "ttnn.scatter"(%5, %127, %128) <{dim = 0, scatter_reduce_type = #ttcore.reduce_type<sum>}>
     : (tensor<1261888xf32>, tensor<6935360xsi32>, tensor<6935360xf32>) -> tensor<1261888xf32>
```

The TILE-layout 1-D intermediate vs. the row-major one the scatter consumes:

```mlir
#ttnn_layout57 = …memref<1x216730x!ttcore.tile<32x32, f32>, #dram>…   // %121: TILE  → 887 MB
#ttnn_layout60 = …memref<1x6935360xf32, #dram>…                       // %128: ROW-MAJOR → 28 MB
```

**Byte math (the 32×):**

| Tensor | shape / layout | bytes |
|---|---|---|
| logical data | `6,935,360 × f32` | `27,741,440` (~28 MB) |
| `%121` TILE `[1, 6935360]` | `216,730 tiles × 32×32 × 4 B` | **`887,726,080`** (~847 MiB) |

`887,726,080 / 27,741,440 = 32.0` exactly — the single real row (height 1) padded to a full 32-row tile. `887,726,080 B` is the exact size of the failing DRAM allocation:

```
TT_FATAL: Out of Memory: Not enough space to allocate 887726080 B DRAM buffer across 12 banks …
(allocated: 969296416 B, free: 101476768 B, largest free block: 55486464 B)
```

(The sibling softmax-denominator scatter flattens `108365 × 8 = 866,920` → `1×27092` tiles ≈ 110 MB, which fits; the `×64` message scatter is the one that overflows.)

### Steps to reproduce the issue

Minimal, PyG-free reproducer of the `reshape → to_layout(row_major) → ttnn.scatter` operand sequence at PubMed dimensions:

```python
import os
os.environ.setdefault("TTXLA_LOGGER_LEVEL", "DEBUG")
os.environ["XLA_STABLEHLO_COMPILE"] = "1"
import torch, torch_xla
import torch_xla.runtime as xr
xr.set_device_type("TT")

torch_xla.set_custom_compile_options({"export_path": "./irs_repro", "export_tensors": True})

E, F, N = 108365, 64, 19717          # PubMed edges+self-loops, heads*out, nodes
dev = torch_xla.device()
index = torch.randint(0, N, (E,), dtype=torch.long, device=dev)
src   = torch.randn(E, F, device=dev)
out   = torch.zeros(N, F, device=dev)
out = out.scatter_add(0, index.unsqueeze(-1).expand(-1, F), src)
torch_xla.sync(wait=True)
```

1. Run the script on a single N300 with a build containing the #8718 fix.
2. Open the exported `./irs_repro/.../ttnn_*.mlir`.
3. Find the `ttnn.reshape` to `[E*F]` that feeds `ttnn.scatter`.
4. Note its layout is `memref<1x…x!ttcore.tile<32x32>>` (TILE) immediately before a `ttnn.to_layout<row_major>` — i.e. a tile-padded 1-D buffer that is discarded for a row-major copy. At PubMed scale this transient is 887 MB and OOMs DRAM in the backward pass of full GATv2 training.

(Originally observed end-to-end: stock `GATv2Conv` (`use_spmm=false`) training on full PubMed — forward/eval runs on device, the first training step OOMs with the `887726080 B` allocation above.)

### Expected behavior

The flattened scatter operand should not be materialized in tile layout with a padded degenerate dim. Either:

- emit the `reshape` result directly in **row-major** (the consumer `to_layout` converts to row-major anyway — fuse/fold it), or
- scatter on the `[E, F]` shape without flattening to a degenerate `[1, E*F]` tile tensor.

Either removes the 32× transient (887 MB → ~28 MB) and lets stock `GATv2Conv` **train** on full PubMed on a single N300.

### Please complete the following environment information

- `pjrt-plugin-tt==1.3.0.dev20260624002726` (tt-mlir `70ff200c…`, tt-metal `7f6364a1`, built 2026-06-23) — contains the #8718 fix
- `torch_xla 2.9.0+git8d31cb3`
- Wormhole b0 (N300), single chip (`mesh_shape: null`), KMD 2.7.0, fw bundle 19.5.0

**Related:** #8714 / #8718 (scatter compile blow-up - fixed; this is the next wall it exposed); #8321 (`grid_sample` tile-padded-constant DRAM OOM, 32× blow-up - sibling *symptom*, different trigger op).

**Comment by igrn256** [comment]
Picking up #8887 (currently unassigned). One design question before I open a PR.

The fix is essentially a sibling of ReshapeRowMajorAdjusting, but for the reshape -> to_layout(row_major) chain rather than reshape -> permute. The idea is to emit the reshape directly in row-major (untiling the input if needed), so the large tile-padded intermediate is never materialized. The following to_layout then folds away via foldIdentityToLayoutOp.

Where would you prefer this transformation to live?

A. A new OpRewritePattern<ReshapeOp> in TTNNMemoryManagement, gated by aggressiveMemorySaving, alongside ReshapeRowMajorAdjusting.
B. A ToLayoutOp canonicalization.

My inclination is A. In the motivating case, the input is tiled, so moving the row-major conversion before the reshape doesn't reduce the number of operations—it only shifts the untiling from the large output to the much smaller input, reducing peak memory rather than performing a canonical simplification.

One related question: the existing sibling uses a fixed 1 GB threshold, so the motivating ~887 MB case doesn't trigger. For this pattern I'm considering using a semantic condition instead:

tiled_volume > row_major_volume

i.e. fire only when tiling actually inflates storage. Does that sound preferable to another absolute threshold?

**Comment by igrn256** [comment]
@St1viG - you filed #8887 and worked around it with SpMM in tenstorrent/tt-blacksmith#611, which leaves the stock `use_spmm: false` path in the tree "until #8887 lands". Quick question on that path: now that SpMM is the working route, is `use_spmm: false` still something you rely on / expect to run, or has it become effectively dormant? Trying to gauge whether the lowering fix still has a live consumer.

@djstefanovicTT - this would be a direct sibling of your `ReshapeRowMajorAdjusting` (#8568): when a reshape's only consumer is a `to_layout` to row-major, emit the reshape in row-major layout directly (untiling the input first) so the 1-D tile-padded transient is never
materialized. On the ticket's IR the reshape result goes ~887 MB -> ~28 MB and the `to_layout` folds away. Working, host-validated (opt + lit, negative check on a rank-1 case), ttsim pending.

If you think this fix makes sense, would you prefer a pass pattern (under `aggressiveMemorySaving`, next to `ReshapeRowMajorAdjusting`) or a `ToLayoutOp` canonicalization? Happy to hold if you'd rather not touch this path right now.

**Comment by djstefanovicTT** [comment]
@igrn256 It makes sense to put this into the TTNNMemoryManagement pass. The minimum threshold memory for some patterns in that pass is added so simple and small cases that do not burden DRAM remain in tile layout to avoid adding unnecessary ops. Here such guard is unnecessary because reshape directly feeds toLayout(RM) operation. I think it would be OK to just check if tiled volume is greater than logical volume. I would also guard it to make sure that the only consumer of reshape is that toLayout op.

**Comment by igrn256** [comment]
Thanks — pass it is.

On the volume check: the motivating reshape is rank-1, and getTilePaddedShape only pads existing dims, so since `E*F` is already a multiple of 32 its tile-padded volume equals its logical volume - `tiled > logical` reads false and skips this case. The 32× is purely a property of the tiled layout (`(d0)->(0,d0)` maps 1-D to `[1, E*F]`, height 1 padded to a full tile), not of the logical shape.

Since the reshape has exactly one consumer and it's `toLayout(RM)`, the intermediate tiled buffer is    dead under the existing structural guards (`hasOneUse` + sole-consumer-toLayout). 
So I'd prefer dropping the volume check and relying on the structural match.

If you'd rather keep an explicit guard across other shapes/ranks, I'd derive it from the physical footprint (tiled memref vs row-major memref) rather than logical volume, so it reflects layout directly. 

Let me know your preference.

**Comment by djstefanovicTT** [comment]
But if it is 1xEF tiled volume is 32xEF? That is the check I was thinking about. I don't see a need to utilize tensors that are tile aligned.

**Comment by igrn256** [comment]
Checked it: getTiledVolume uses getTilePaddedShape(getShape()), and here getShape() is rank-1 [6935360] -> stays [6935360] (already ×32-aligned), so tiledVol == rmVol and the check is false. But getShardShape() reads the layout and 
returns [1, 216730] (× 4096 B/tile ≈ 887 MB vs 28 MB row-major) - so the 32× is visible there. Happy to guard on the layout footprint via getShardShape() rather than getTilePaddedShape(getShape()), or keep it purely structural (sole consumer is toLayout(RM)). 
Which do you prefer?

**Comment by djstefanovicTT** [comment]
I would prefer if there is a guard.

**Comment by igrn256** [comment]
Hi, I'd like to pick this one up. Could you assign it to me?

I have a candidate fix locally. The issue seems to be caused by the flattened `[E*F]`
operand being materialized in `TILE` layout as `[1, E*F]`. The degenerate height of 1
is padded to a full 32-row tile, inflating the buffer footprint significantly
(~847 MiB vs ~28 MiB in my repro).

The following `to_layout` converts it back to row-major before `ttnn.scatter` consumes
it, so the tiled representation is only a transient allocation.

My current approach is to move the row-major conversion before the reshape, so the
reshape emits row-major directly and the padded `TILE` buffer is never materialized. I
implemented this as a new pattern in `TTNNMemoryManagement.cpp`, modelled after the
existing `ReshapeRowMajorAdjusting` and `PermuteRowMajorAdjusting` patterns.

I have lit tests and full `check-ttmlir` passing locally, but I have not verified it
on device yet, as I don't have TT hardware.

Before opening a PR, I wanted to clarify two design points:

1. Should this stay a narrow optimization for the `reshape -> to_layout(row-major)`
   pattern, or should it become a more general policy for degenerate dimensions (any
   tiled tensor whose padded size far exceeds its logical size)?

2. The neighbouring patterns use a fixed 1 GiB threshold. For this case I instead
   compare the `TILE` and row-major sizes via `getShardSizeInBytes()`, since the
   row-major conversion already follows immediately. Does this match the intended
   approach, or should this also use the threshold-based guard?

**Comment by djstefanovicTT** [comment]
>Should this stay a narrow optimization for the reshape -> to_layout(row-major)
pattern, or should it become a more general policy for degenerate dimensions (any
tiled tensor whose padded size far exceeds its logical size)?

It should stay narrow. Only thing that I would consider generalizing is that you can make it templated pattern and for now to accept reshape.

>The neighbouring patterns use a fixed 1 GiB threshold. For this case I instead
compare the TILE and row-major sizes via getShardSizeInBytes(), since the
row-major conversion already follows immediately. Does this match the intended
approach, or should this also use the threshold-based guard?

I don't think that there should be any threshold here.
